### ISSUE TITLE
Call to undefined function wc_get_chosen_shipping_method_ids()

### ISSUE THREAD

**Issue by txl518** [woothemes/woocommerce#11959]
## EXPLANATION OF THE ISSUE

Call to undefined function wc_get_chosen_shipping_method_ids() in class-wc-customer.php this is only affecting variable products. I added require_once seems to fix the issue for now.

**Error log:**

2016/09/22 02:52:05 [error] 2122#2122: *486 FastCGI sent in stderr: "PHP message: PHP Fatal error:  Call to undefined function wc_get_chosen_shipping_method_ids() in /var/www/html/wp-content/plugins/woocommerce/includes/class-wc-customer.php on line 333
PHP message: PHP Stack trace:
PHP message: PHP   1. {main}() /var/www/html/wp-admin/admin-post.php:0
PHP message: PHP   2. do_action() /var/www/html/wp-admin/admin-post.php:69
PHP message: PHP   3. call_user_func_array:{/var/www/html/wp-includes/plugin.php:524}() /var/www/html/wp-includes/plugin.php:524
PHP message: PHP   4. FUE_Admin_Actions::send_queue_item() /var/www/html/wp-includes/plugin.php:524
PHP message: PHP   5. FUE_Sending_Mailer->send_queue_item() /var/www/html/wp-content/plugins/woocommerce-follow-up-emails/includes/class-fue-admin-actions.php:1406
PHP message: PHP   6. FUE_Sending_Mailer->process_variable_replacements() /var/www/html/wp-content/plugins/woocommerce-follow-up-emails/includes/sending/class-fue-sending-mailer.php:202
PHP message: PHP   7. do_action() /var/www/html/wp-content/plugins/woocommerce-follow-up-emails/includes/sending/class-fue-sending-mailer.php:871
PHP message: PHP   8. call_user_func_array:{/var/www/html/wp-includes/plugin.php:524}() /var/www/html/wp-includes/plugin.php:524
PHP message: PHP   9. FUE_Addon_Woocommerce_Mailer->register_email_variable_replacements() /var/www/html/wp-includes/plugin.php:524
PHP message: PHP  10. FUE_Addon_Woocommerce_Mailer->add_variable_replacements() /var/www/html/wp-content/plugins/woocommerce-follow-up-emails/includes/addons/class-fue-addon-woocommerce-mailer.php:377
PHP message: PHP  11. FUE_Addon_Woocommerce_Mailer->get_replacement_data() /var/www/html/wp-content/plugins/woocommerce-follow-up-emails/includes/addons/class-fue-addon-woocommerce-mailer.php:502
PHP message: PHP  12. FUE_Addon_Woocommerce_Mailer->get_replacement_data_for_product() /var/www/html/wp-content/plugins/woocommerce-follow-up-emails/includes/addons/class-fue-addon-woocommerce-mailer.php:704
PHP message: PHP  13. WC_P

## STEPS TO REPRODUCE THE ISSUE

1. Vanilla Woocommerce site 
2. Install follow up email plugins and 
2. Add a variable product with variations ( or add simple product to see)
3. put item in shopping cart go to check out page enter email and etc for follow up email to capture
4. click send it now.
5. see error log

## SYSTEM STATUS REPORT
`
### WordPress Environment ###

Home URL: http://1.1.1.1
Site URL: http://1.1.1.1
WC Version: 2.6.4
Log Directory Writable: ✔
WP Version: 4.6.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US

### Server Environment ###

Server Info: nginx/1.10.1
PHP Version: 5.6.25
PHP Post Max Size: 2 GB
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 7.19.7
NSS/3.21 Basic ECC

SUHOSIN Installed: –
MySQL Version: 5.6.33
Max Upload Size: 2 GB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ❌ Your server does not have the SoapClient class enabled - some gateway plugins which use SOAP may not work as expected.
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 2.6.4
: 
woocommerce_sessions: ✔
woocommerce_api_keys: ✔
woocommerce_attribute_taxonomies: ✔
woocommerce_downloadable_product_permissions: ✔
woocommerce_order_items: ✔
woocommerce_order_itemmeta: ✔
woocommerce_tax_rates: ✔
woocommerce_tax_rate_locations: ✔
woocommerce_shipping_zones: ✔
woocommerce_shipping_zone_locations: ✔
woocommerce_shipping_zone_methods: ✔
woocommerce_payment_tokens: ✔
woocommerce_payment_tokenmeta: ✔
MaxMind GeoIP Database: ❌ The MaxMind GeoIP Database does not exist - Geolocation will not function. You can download and install it manually from http://dev.maxmind.com/geoip/legacy/geolite/ to the path: . Scroll down to \"Downloads\" and download the \"Binary / gzip\" file next to \"GeoLite Country\"

### Active Plugins (2) ###

Follow-Up Emails: by WooThemes – 4.4.16
WooCommerce: by WooThemes – 2.6.4

### Settings ###

Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2

### API ###

API Enabled: ✔

### WC Pages ###

Shop Base: #4 - /shop/
Cart: #5 - /cart/
Checkout: #6 - /checkout/
My Account: #7 - /my-account/

### Taxonomies ###

Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

### Theme ###

Name: Twenty Sixteen
Version: 1.3
Author URL: https://wordpress.org/
Child Theme: ❌ – If you're modifying WooCommerce on a parent theme you didn't build personally
then we recommend using a child theme. See: How to create a child theme

WooCommerce Support: ✔

**Comment by mikejolley** [comment]
Exists here https://github.com/woothemes/woocommerce/blob/481471a64cccc97ffd084166f17ecf399dbd4ff3/includes/wc-cart-functions.php#L375 No issues my side.

Might be FUE - raise a ticket on support.woothemes.com with steps to reproduce so the dev can look into this properly.

**Comment by mikejolley** [comment]
Found an existing report of this so passing to the ninja handling it so FUE can be looked at further if needed.

**Comment by saulfautley** [comment]
This error is caused by Follow-up Emails instantiating the `WC_Customer` class, which seems to only be intended to be used on the frontend. 
```php
# woocommerce-follow-up-emails/includes/addons/class-fue-addon-woocommerce-cart.php:398

WC()->customer = new WC_Customer();
```

This makes WooCommerce think that it's working on the frontend and that all its frontend dependencies have been loaded. So when `WC_Product_Variable::get_price_html()` is called by Follow-up Emails, `WC_Tax::get_tax_location()` is called which checks for `WC_Customer`.
```php
# woocommerce/includes/class-wc-tax.php:376

public static function get_tax_location( $tax_class = '' ) {
	$location = array();

	if ( ! empty( WC()->customer ) ) {
		$location = WC()->customer->get_taxable_address();
	} ...
```

The `WC_Customer::get_taxable_address()` method tries to call `wc_get_chosen_shipping_method_ids()`, which - you guessed it - has not been loaded.

This could be resolved by Follow-up Emails making sure it loads the `wc-cart-functions.php` file where necessary, or if WooCommerce wanted to allow `WC_Customer` to be used when not on the frontend, then `WC_Tax::get_tax_location()` or `WC_Customer::get_taxable_address()` would need to be refactored a bit.

**Comment by Danw33** [comment]
We're seeing this issue also with the FUE plugin, partial stack trace with paths shortened:

```
Fatal error: Uncaught Error: Call to undefined function wc_get_chosen_shipping_method_ids() in [...]/wp-content/plugins/woocommerce/includes/class-wc-customer.php:333
Stack trace:
#0 [...]/wp-content/plugins/woocommerce/includes/class-wc-tax.php(377): WC_Customer->get_taxable_address()
#1 [...]/wp-content/plugins/woocommerce/includes/class-wc-tax.php(397): WC_Tax::get_tax_location('')
#2 [...]/wp-content/plugins/woocommerce/includes/class-wc-cart.php(1218): WC_Tax::get_rates('')
#3 [...]/wp-content/plugins/woocommerce/includes/class-wc-cart.php(261): WC_Cart->calculate_totals()
#4 [...]/wp-content/plugins/woocommerce/includes/class-wc-cart.php(159): WC_Cart->get_cart_from_session()
#5 [...]/wp-content/plugins/woocommerce-follow-up-emails/includes/addons/class-fue-addo in [...]/wp-content/plugins/woocommerce/includes/class-wc-customer.php on line 333
```

@saulfautley Looks like your onto something there. I wonder if there is a fix in the pipeline?

**Comment by KloonDuino** [comment]
At the moment this error is still there.

**Comment by Danw33** [comment]
@KloonDuino Yep, they've not released v5.0 yet that fixes the issue. 😞 

Your best bet is to get in touch with Woo and ask them for the latest 5.0-RC which has the patch in 👍
